File: python/src/bellman_ford.py
"""Bellman-Ford Algorithm Implementation

Finds shortest paths from a source node to all other nodes, even with negative edge weights.
Can also detect negative cycles.

Time Complexity: O(VE)
Space Complexity: O(V)
"""

from typing import Dict, List, Tuple, Optional, Set


def bellman_ford(
    graph: Dict[str, Dict[str, float]], 
    start: str
) -> Tuple[Dict[str, float], Dict[str, Optional[str]]]:
    """
    Bellman-Ford algorithm for shortest paths with negative weights.
    
    Args:
        graph: Adjacency list representation {node: {neighbor: weight}}
        start: Starting node
        
    Returns:
        Tuple of (distances, predecessors)
        - distances: Dict mapping each node to its shortest distance from start
        - predecessors: Dict mapping each node to its predecessor in the shortest path
        
    Raises:
        ValueError: If a negative cycle is detected
        
    Example:
        >>> graph = {
        ...     'A': {'B': -1, 'C': 4},
        ...     'B': {'C': 3, 'D': 2},
        ...     'C': {},
        ...     'D': {'B': 1, 'C': 5}
        ... }
        >>> distances, predecessors = bellman_ford(graph, 'A')
        >>> distances['C']
        2
    """
    # Initialize distances and predecessors
    distances: Dict[str, float] = {node: float('inf') for node in graph}
    distances[start] = 0
    predecessors: Dict[str, Optional[str]] = {node: None for node in graph}
    
    # Get all nodes
    nodes = list(graph.keys())
    num_nodes = len(nodes)
    
    # Relax edges V-1 times
    for _ in range(num_nodes - 1):
        updated = False
        
        for node in nodes:
            if distances[node] == float('inf'):
                continue
                
            for neighbor, weight in graph[node].items():
                new_distance = distances[node] + weight
                
                if new_distance < distances[neighbor]:
                    distances[neighbor] = new_distance
                    predecessors[neighbor] = node
                    updated = True
        
        # Early termination if no updates
        if not updated:
            break
    
    # Check for negative cycles
    for node in nodes:
        if distances[node] == float('inf'):
            continue
            
        for neighbor, weight in graph[node].items():
            if distances[node] + weight < distances[neighbor]:
                raise ValueError("Graph contains a negative cycle")
    
    return distances, predecessors


def detect_negative_cycle(graph: Dict[str, Dict[str, float]]) -> Optional[List[str]]:
    """
    Detect if a graph contains a negative cycle and return it.
    
    Args:
        graph: Adjacency list representation
        
    Returns:
        List of nodes forming a negative cycle, or None if no cycle exists
        
    Example:
        >>> graph = {
        ...     'A': {'B': 1},
        ...     'B': {'C': -3},
        ...     'C': {'A': 1}
        ... }
        >>> cycle = detect_negative_cycle(graph)
        >>> cycle
        ['A', 'B', 'C', 'A']
    """
    # Initialize distances and predecessors
    nodes = list(graph.keys())
    distances: Dict[str, float] = {node: float('inf') for node in nodes}
    predecessors: Dict[str, Optional[str]] = {node: None for node in nodes}
    
    # Use first node as arbitrary start
    if not nodes:
        return None
    
    start = nodes[0]
    distances[start] = 0
    
    num_nodes = len(nodes)
    
    # Relax edges V-1 times
    for _ in range(num_nodes - 1):
        for node in nodes:
            if distances[node] == float('inf'):
                continue
                
            for neighbor, weight in graph[node].items():
                new_distance = distances[node] + weight
                
                if new_distance < distances[neighbor]:
                    distances[neighbor] = new_distance
                    predecessors[neighbor] = node
    
    # Check for negative cycles and find a node in the cycle
    cycle_node: Optional[str] = None
    
    for node in nodes:
        if distances[node] == float('inf'):
            continue
            
        for neighbor, weight in graph[node].items():
            if distances[node] + weight < distances[neighbor]:
                cycle_node = neighbor
                break
        
        if cycle_node:
            break
    
    if not cycle_node:
        return None
    
    # Trace back to find the cycle
    # Move back V times to ensure we're in the cycle
    for _ in range(num_nodes):
        cycle_node = predecessors[cycle_node]
    
    # Now trace the cycle
    cycle: List[str] = []
    current = cycle_node
    
    while True:
        cycle.append(current)
        current = predecessors[current]
        
        if current == cycle_node:
            cycle.append(current)
            break
    
    cycle.reverse()
    return cycle


def bellman_ford_path(
    graph: Dict[str, Dict[str, float]], 
    start: str, 
    end: str
) -> Tuple[float, List[str]]:
    """
    Find the shortest path between two specific nodes using Bellman-Ford.
    
    Args:
        graph: Adjacency list representation
        start: Starting node
        end: Target node
        
    Returns:
        Tuple of (distance, path)
        
    Raises:
        ValueError: If no path exists or negative cycle detected
    """
    distances, predecessors = bellman_ford(graph, start)
    
    # Check if path exists
    if distances[end] == float('inf'):
        raise ValueError(f"No path exists from {start} to {end}")
    
    # Reconstruct path
    path: List[str] = []
    current: Optional[str] = end
    
    while current is not None:
        path.append(current)
        current = predecessors[current]
    
    path.reverse()
    
    return distances[end], path

File: go/pkg/bellmanford/bellman_ford.go
package bellmanford

import (
	"fmt"
	"math"
)

// Graph represents a graph where edges have weights
type Graph map[string]map[string]float64

// Result contains the shortest distances and predecessors
type Result struct {
	Distances    map[string]float64
	Predecessors map[string]*string
}

// CheckNegativeCycleResult contains the cycle if found
type CheckNegativeCycleResult struct {
	HasCycle bool
	Cycle    []string
}

// Run executes the Bellman-Ford algorithm
// Returns error if a negative cycle is detected
func Run(graph Graph, start string) (*Result, error) {
	distances := make(map[string]float64)
	predecessors := make(map[string]*string)

	// Initialize
	for node := range graph {
		distances[node] = math.Inf(1)
		predecessors[node] = nil
	}
	distances[start] = 0

	// Get all nodes for iteration
	nodes := make([]string, 0, len(graph))
	for node := range graph {
		nodes = append(nodes, node)
	}

	// Relax edges |V| - 1 times
	for i := 0; i < len(nodes)-1; i++ {
		updated := false
		for u := range graph {
			if math.IsInf(distances[u], 1) {
				continue
			}
			for v, weight := range graph[u] {
				if distances[u]+weight < distances[v] {
					distances[v] = distances[u] + weight
					curr := u
					predecessors[v] = &curr
					updated = true
				}
			}
		}
		if !updated {
			break
		}
	}

	// Check for negative cycles
	for u := range graph {
		if math.IsInf(distances[u], 1) {
			continue
		}
		for v, weight := range graph[u] {
			if distances[u]+weight < distances[v] {
				return nil, fmt.Errorf("negative cycle detected")
			}
		}
	}

	return &Result{
		Distances:    distances,
		Predecessors: predecessors,
	}, nil
}

// DetectNegativeCycle checks for negative cycles and returns one if found
func DetectNegativeCycle(graph Graph) []string {
	distances := make(map[string]float64)
	predecessors := make(map[string]*string)

	// Identify all nodes
	nodes := make([]string, 0, len(graph))
	for node := range graph {
		nodes = append(nodes, node)
	}

	if len(nodes) == 0 {
		return nil
	}

	// Initialize with arbitrary start node (or all 0 if we want to find any cycle reachable from anywhere)
	// For cycle detection in disconnected components, we'd typically add a dummy source connected to all nodes with 0 weight.
	// Here we'll just try to relax from a source. If the graph is connected, it works.
	// A more robust way for ANY cycle is to set all distances to 0.
	for _, node := range nodes {
		distances[node] = 0
		predecessors[node] = nil
	}

	// Relax |V| - 1 times
	for i := 0; i < len(nodes)-1; i++ {
		for u := range graph {
			for v, weight := range graph[u] {
				if distances[u]+weight < distances[v] {
					distances[v] = distances[u] + weight
					curr := u
					predecessors[v] = &curr
				}
			}
		}
	}

	// Check for cycle
	var cycleNode *string
	for u := range graph {
		for v, weight := range graph[u] {
			if distances[u]+weight < distances[v] {
				// Cycle detected
				node := v
				cycleNode = &node
				// Retrace back to ensure we are in the cycle
				for i := 0; i < len(nodes); i++ {
					if predecessors[*cycleNode] != nil {
						node := *predecessors[*cycleNode]
						cycleNode = &node
					}
				}
				goto Found
			}
		}
	}
	return nil

Found:
	// Trace the cycle
	cycle := make([]string, 0)
	curr := *cycleNode
	for {
		cycle = append(cycle, curr)
		if predecessors[curr] != nil {
			curr = *predecessors[curr]
			if curr == *cycleNode && len(cycle) > 0 {
				cycle = append(cycle, curr)
				break
			}
		} else {
			break // Should not happen in a cycle
		}
	}

	// Reverse to get correct order
	for i, j := 0, len(cycle)-1; i < j; i, j = i+1, j-1 {
		cycle[i], cycle[j] = cycle[j], cycle[i]
	}

	return cycle
}
